From e236b3d721979cb6c2f4aede8961b12e82da1809 Mon Sep 17 00:00:00 2001
From: Mythri P K <mythri.p.k@intel.com>
Date: Tue, 22 Apr 2014 18:17:40 +0530
Subject: [PATCH 09/39] ASoC: soc-fw: Add the pvt load call back

Issue: GMINL-3679

Provide a way for the registered component driver to have
a callback to fill the private data passed from the
fw.

Change-Id: I4ebc45eac251a27b39d8d5e313ff68f49ddd04ff
Signed-off-by: Mythri P K <mythri.p.k@intel.com>
---
 include/sound/soc-fw.h |  7 +++++++
 sound/soc/soc-fw.c     | 17 +++++++++++++++++
 2 files changed, 24 insertions(+)

diff --git a/include/sound/soc-fw.h b/include/sound/soc-fw.h
index 8d5d4e1..3928a0b 100644
--- a/include/sound/soc-fw.h
+++ b/include/sound/soc-fw.h
@@ -55,6 +55,9 @@ struct snd_soc_fw_codec_ops {
 	int (*vendor_load)(struct snd_soc_codec *, struct snd_soc_fw_hdr *);
 	int (*vendor_unload)(struct snd_soc_codec *, struct snd_soc_fw_hdr *);
 
+
+	int (*pvt_load)(struct snd_soc_codec *, u32, unsigned long, unsigned long);
+
 	/* completion - called at completion of firmware loading */
 	void (*complete)(struct snd_soc_codec *);
 
@@ -78,6 +81,8 @@ struct snd_soc_fw_platform_ops {
 	int (*vendor_load)(struct snd_soc_platform *, struct snd_soc_fw_hdr *);
 	int (*vendor_unload)(struct snd_soc_platform *, struct snd_soc_fw_hdr *);
 
+	int (*pvt_load)(struct snd_soc_platform *, u32, unsigned long, unsigned long);
+
 	/* completion - called at completion of firmware loading */
 	void (*complete)(struct snd_soc_platform *);
 
@@ -102,6 +107,8 @@ struct snd_soc_fw_card_ops {
 	int (*vendor_load)(struct snd_soc_card *, struct snd_soc_fw_hdr *);
 	int (*vendor_unload)(struct snd_soc_card *, struct snd_soc_fw_hdr *);
 
+	int (*pvt_load)(struct snd_soc_card *, u32, unsigned long, unsigned long);
+
 	/* completion */
 	void (*complete)(struct snd_soc_card *);
 
diff --git a/sound/soc/soc-fw.c b/sound/soc/soc-fw.c
index 8185d41..e9ab2d1 100644
--- a/sound/soc/soc-fw.c
+++ b/sound/soc/soc-fw.c
@@ -362,6 +362,23 @@ static int soc_fw_init_kcontrol(struct soc_fw *sfw, struct snd_kcontrol_new *k)
 	return 0;
 }
 
+/* optionally pass private data to be habdled by component driver. */
+static int soc_fw_init_pvt_data(struct soc_fw *sfw, u32 io_type, unsigned long sm,
+			unsigned long mc)
+{
+	if (sfw->codec && sfw->codec_ops && sfw->codec_ops->pvt_load)
+		return sfw->codec_ops->pvt_load(sfw->codec, io_type, sm, mc);
+
+	if (sfw->platform && sfw->platform_ops && sfw->platform_ops->pvt_load)
+		return sfw->platform_ops->pvt_load(sfw->platform, io_type, sm, mc);
+
+	if (sfw->card && sfw->card_ops && sfw->card_ops->pvt_load)
+		return sfw->card_ops->pvt_load(sfw->card, io_type, sm, mc);
+
+	dev_dbg(sfw->dev, "ASoC: no handler specified for pvt data copy\n");
+	return 0;
+}
+
 static int soc_fw_create_tlv(struct soc_fw *sfw, struct snd_kcontrol_new *kc,
 	u32 tlv_size)
 {
-- 
1.9.1

